server/datastore/mysql/software.go

```diff
@@ -2877,301 +3144,601 @@ last_vpp_install AS (
+		if len(softwareTitleIds) > 0 {
+			replacements = append(replacements,
+				// For software installers
+				`
+				SELECT
+					software_titles.id,
+					software_titles.name,
+					software_titles.source AS source,
+					software_installers.id AS installer_id,
+					software_installers.self_service AS package_self_service,
+					software_installers.filename AS package_name,
+					software_installers.version AS package_version,
+					software_installers.platform as package_platform,
+					GROUP_CONCAT(software.id) AS software_id_list,
+					GROUP_CONCAT(software.source) AS software_source_list,
+					GROUP_CONCAT(software.version) AS version_list,
+					GROUP_CONCAT(software.bundle_identifier) AS bundle_identifier_list,
+					NULL AS vpp_app_adam_id_list,
+					NULL AS vpp_app_version_list,
+					NULL AS vpp_app_platform_list,
+					NULL AS vpp_app_icon_url_list,
+					NULL AS vpp_app_self_service_list
+			`, softwareVulnerableJoin, `
+				GROUP BY
+					software_titles.id,
+					software_titles.name,
+					software_titles.source,
+					software_installers.id,
+					software_installers.self_service,
+					software_installers.filename,
+					software_installers.version,
+					software_installers.platform
+			`)
+		}
+		if len(vppAdamIDs) > 0 {
+			replacements = append(replacements,
+				// For vpp apps
+				`
+				SELECT
+					software_titles.id,
+					software_titles.name,
+					software_titles.source AS source,
+					software_installers.id AS installer_id,
+					software_installers.self_service AS package_self_service,
+					software_installers.filename AS package_name,
+					software_installers.version AS package_version,
+					software_installers.platform as package_platform,
+					NULL AS software_id_list,
+					NULL AS software_source_list,
+					NULL AS version_list,
+					NULL AS bundle_identifier_list,
+					GROUP_CONCAT(vpp_apps.adam_id) AS vpp_app_adam_id_list,
+					GROUP_CONCAT(vpp_apps.latest_version) AS vpp_app_version_list,
+					GROUP_CONCAT(vpp_apps.platform) as vpp_app_platform_list,
+					GROUP_CONCAT(vpp_apps.icon_url) AS vpp_app_icon_url_list,
+					GROUP_CONCAT(vpp_apps_teams.self_service) AS vpp_app_self_service_list
+			`, vppVulnerableJoin, `
+				GROUP BY
+					software_titles.id,
+					software_titles.name,
+					software_titles.source,
+					software_installers.id,
+					software_installers.self_service,
+					software_installers.filename,
+					software_installers.version,
+					software_installers.platform
+			`)
+		}
+		stmt = fmt.Sprintf(stmt, replacements...)
+		stmt = fmt.Sprintf("SELECT * FROM (%s) AS combined_results", stmt)
+		stmt, _ = appendListOptionsToSQL(stmt, &opts.ListOptions)
+
+		stmt = ds.reader(ctx).Rebind(stmt)
+		if err := sqlx.SelectContext(ctx, ds.reader(ctx), &hostSoftwareList, stmt, args...); err != nil {
+			return nil, nil, ctxerr.Wrap(ctx, err, "list host software")
+		}
+
+		// collect install paths by software.id
+		installedPaths, err := ds.getHostSoftwareInstalledPaths(ctx, host.ID)
```

Comment:
We could probably add an optional slice of software ids to `getHostSoftwareInstalledPaths`. Where in this usage we could filter by software id, where as everywhere else in the code, we could pass an empty slice and it would retain it's original functionality.